### ISSUE TITLE
object attribute mutator methods should be documented - maybe in Perl6/doc

### ISSUE THREAD

**Issue by ronaldxs** [perl6/doc#136]
In S12 (Objects) in the section on attributes - http://design.perl6.org/S12.html#Attributes - there is a sentence that reads:  "You may write your own accessors to override any or all of the autogenerated ones.".  The same is supposed to be true about mutators for rw attributes but that does not seem to be documented anywhere.  After a discussion with moritz++ on irc, moritz felt that a Synopsis does not need to spell everything out and while custom mutators should be documented, that documentation belonged on Perl6/doc.  I don't really see where custom accessors are documented on Perl6/doc but, like accessors, custom mutators are supposed to be part of the Perl 6 language and that should be documented somewhere.

There is a work in progress roast test suite for custom object mutators here: https://github.com/perl6/roast/blob/master/S12-attributes/mutators.t

**Comment by LLFourn** [comment]
There is a mention of it: http://doc.perl6.org/language/containers#Scalar_containers

But there should be some mention of it in the context of objects on the classes page. To me "mutators" would fit into a tutorial like thing is it is just the combination of Proxy objects and custom rw accessors.

**Comment by ronaldxs** [comment]
For "Scalar Containers" they give a trivial mutator example that isn't really a custom mutator like in S12-attributes/mutators.t.

```
class A {
    has $.attr is rw;
}

is equivalent to

class A {
    has $!attr;
    method attr() is rw { $!attr }
}
```
This is a minimal mutator and somewhere there needs to be documentation about object attribute mutators that can do a lot more than this.

**Comment by LLFourn** [comment]
My suggestion is that a fuller mutator example involving a Proxy like that test belongs in [classtut](http://docs.perl6.org/language/classtut) while in the [objects doc](http://doc.perl6.org/language/objects) we mention that methods can also take `is rw` and this allows you to make custom accessors.

**Comment by AlexDaniel** [comment]
Perhaps the fact that this issue is one year old is scaring away some contributors :)
Marking as LHF.

**Comment by JJ** [comment]
Two years old and counting... I'll get it and see what I can do about it.

**Comment by ronaldxs** [comment]
There is an example that may be helpful towards the end (section labeled "Customizing") of the Day 2 Advent Calendar 2017 article: https://perl6advent.wordpress.com/2017/12/02/perl-6-sigils-variables-and-containers

**Comment by JJ** [comment]
OK, I'll run this through you before inserting it into the documentation. First thing is that, in general, introducing other concepts, such as `Proxy`, in the documentation of things like this is generally frowned upon. It might be OK for a tutorial, but not for the object documentation itself. So instead of a FETCH and a STORE, I've created two multi methods, one for fetch and another for store:

```
class Journey {
    has $.origin;
    has $.destination;
    has @.travelers;
    has Str $!notes = "==Start==\n";

    multi method notes() { $!notes };
    multi method notes( Str $note ) { $!notes ~= "\n" ~ $note };
}

my $trip = Journey.new( :origin<Here>, :destination<There>,
			travelers => <þor Freya> );

say "\t→" ~ $trip.notes();
$trip.notes("First steps");
say "\t→" ~ $trip.notes();
```

I'm not sure this is exactly what you are looking for; but it's the best I can do. Any help will be appreciated.

**Comment by ronaldxs** [comment]
I am sick and incapacitated for the forseeable futire.  Will try to come up with some answer by the end of March.

**Comment by JJ** [comment]
Sorry to hear that. Get well soon.

**Comment by ronaldxs** [comment]
Python, Ruby, C# and Delphi all allow 'setters' for attributes that overload the assignment operator so people from other languages may wonder about the feature which seems to require a `Proxy` in p6.  See PR #1874 .

**Comment by JJ** [comment]
Glad to hear you're back. :-)

**Comment by ronaldxs** [comment]
Closed by PR #1874.
